### PR TITLE
INTERCEPTS overflow emulation from Chocolate Doom

### PR DESCRIPTION
**Pull request by rfomin**
I didn't mean that I want to do it but oh well :smile:
1. I coudn't find demos with INTERCEPTS overflow that listed here: [[doomworld]](https://www.doomworld.com/forum/topic/35214-spechits-reject-and-intercepts-overflow-lists/) Do you know archive with these kind of demos?
2. Should we implement menu or config entries for various overflow emulations?

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
Thank you very much for doing this anyway! :grin: 

There is the "odd demos" section on Classic Doom where you can find at least two demos that feature this glitch:
http://classicdoom.com/odddemos.htm#manorbug

File: Source/p_maputl.c [hunk 1]
```diff
@@ -583,6 +593,127 @@ boolean P_TraverseIntercepts(traverser_t func, fixed_t maxfrac)
...
+    }
+}
+
+// Emulate overruns of the intercepts[] array.
+
+static void InterceptsOverrun(int num_intercepts, intercept_t *intercept)
+{
+    int location;
+
+    if (num_intercepts <= MAXINTERCEPTS_ORIGINAL)
```
Comment (fabiangreffrath) [inline Source/p_maputl.c:696]: We need a check for demo_version here!

**Review by fabiangreffrath — APPROVED** [review]

**Comment by fabiangreffrath** [comment]
> 2. Should we implement menu or config entries for various overflow emulations?

"In Woof!, compatibility is not an option."

image

**Comment by rfomin** [comment]
> If you need more intercept overflow demos, KD should have some on the discord, he has been keeping a list of examples.

Thanks, but now I'm more intersted in donut overflow and the playingame overflow. I have never seen these in action.

**Comment by fabiangreffrath** [comment]
The latter isn't even emulated by Choco, so maybe not *that* relevant.

**Comment by rfomin** [comment]
> > The latter isn't even emulated by Choco, so maybe not _that_ relevant.
> 
> Apparently, I was wrong. It is just a bit... simplified: [chocolate-doom/chocolate-doom@7994c9e](https://github.com/chocolate-doom/chocolate-doom/commit/7994c9e7e6d8a6f68c8841c65db3f981d1c3ccf2)

There is an article and wad/demo: http://www.doom2.net/doom2/research/runningbody.html I guess we should add this.

**Comment by rfomin** [comment]
Bug with test wad for donut overflow: https://sourceforge.net/p/chocolate-doom/bugs/58/ No crash in Woof.

**Comment by fabiangreffrath** [comment]
The playeringame overflow does already seem to work as intended on my system. I wonder why it doesn't crash when attempting to dereference the -1st element of the playeringame array?

I am tempted to add a check for `players[3].didsecret` as mentioned in PrBoom+'s code somehow...

**Comment by fabiangreffrath** [comment]
> Bug with test wad for donut overflow: https://sourceforge.net/p/chocolate-doom/bugs/58/ No crash in Woof.

Indeed! It seems that Boom/MBF has already taken some security measures to this.

**Comment by fabiangreffrath** [comment]
> There is an article and wad/demo: http://www.doom2.net/doom2/research/runningbody.html I guess we should add this.

Would you take care of that?

**Comment by rfomin** [comment]
> > There is an article and wad/demo: http://www.doom2.net/doom2/research/runningbody.html I guess we should add this.
> 
> Would you take care of that?

I don't know, is it necessary? Only if we need the ability to fix this error (disable emulation).

**Comment by fabiangreffrath** [comment]
> I don't know, is it necessary? Only if we need the ability to fix this bug (disable emulation).

I don't think we need a switch for this, just catch the array overflow. Like this:

```patch
--- a/Source/p_mobj.c
+++ b/Source/p_mobj.c
@@ -1031,7 +1031,7 @@ void P_SpawnPlayer (mapthing_t* mthing)

   // not playing?

-  if (!playeringame[mthing->type-1])
+  if (mthing->type == 0 || !playeringame[mthing->type-1]) // [FG] handle overflow
     return;

   p = &players[mthing->type-1];

```

**Comment by rfomin** [comment]
> I don't think we need a switch for this, just catch the array overflow.

Sure, I agree.

**Pull request merged**